common4j/src/main/com/microsoft/identity/common/java/crypto/key/KeyUtil.java

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.crypto.key;
+
+import com.microsoft.identity.common.java.exception.ClientException;
+import com.microsoft.identity.common.java.logging.Logger;
+import com.microsoft.identity.common.java.util.Base64;
+
+import java.security.InvalidKeyException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+import lombok.NonNull;
+
+import static com.microsoft.identity.common.java.AuthenticationConstants.ENCODING_UTF8;
+
+public class KeyUtil {
+
+    private static final Object TAG = KeyUtil.class.getSimpleName();
+
+    /**
+     * A string to return when the this class fails to derive thumbprint.
+     */
+    public static final String UNKNOWN_THUMBPRINT = "UNKNOWN_THUMBPRINT";
+
+    /**
+     * HMac key hashing algorithm.
+     */
+    public static final String HMAC_KEY_HASH_ALGORITHM = "SHA-256";
+
+    /**
+     * HMac algorithm (for {@link Mac#getInstance(String)})
+     */
+    public static final String HMAC_ALGORITHM = "HmacSHA256";
+
+    /**
+     * HMac Key spec algorithm.
+     */
+    private static final String HMAC_KEYSPEC_ALGORITHM = "AES";
+
+
+    /**
+     * Derive a thumbprint from the given ISecretKeyLoader.
+     *
+     * @param keyLoader ISecretKeyLoader to obtain the key (calculate the thumbprint from).
+     * @return a thumbprint. Will return {@link KeyUtil#UNKNOWN_THUMBPRINT} if it fails to derived one.
+     */
+    public static String getKeyThumbPrint(final @NonNull ISecretKeyLoader keyLoader) {
+        final String methodName = ":getKeyThumbPrint";
+        try {
+            return getKeyThumbPrint(keyLoader.getKey());
```

Comment:
Should this thumbprint mechanism be part of the secretKeyLoader instead of not?

Reply:
I see this as a util function. 

SecretKeyLoader is just a way to load secret key.